File: examples/graphql-spring-boot/src/main/java/org/neo4j/graphql/examples/graphqlspringboot/datafetcher/AdditionalDataFetcher.java
package org.neo4j.graphql.examples.graphqlspringboot.datafetcher;

import com.fasterxml.jackson.annotation.JsonProperty;
import graphql.schema.DataFetchingEnvironment;
import org.springframework.graphql.data.method.annotation.QueryMapping;
import org.springframework.graphql.data.method.annotation.SchemaMapping;
import org.springframework.stereotype.Controller;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;

@Controller
class AdditionalDataFetcher {

    @SchemaMapping(typeName = "Movie", field = "bar")
    public String bar() {
        return "foo";
    }

    @SchemaMapping(typeName = "Movie", field = "javaData")
    public List<JavaData> javaData(DataFetchingEnvironment env) {
        //noinspection unchecked
        Object title = ((Map<String, Object>) Objects.requireNonNull(env.getSource())).get("title");
        return Collections.singletonList(new JavaData("test " + title));
    }

    @QueryMapping
    public String other() {
        return "other";
    }

    public static class JavaData {
        @JsonProperty("name")
        public String name;

        public JavaData(String name) {
            this.name = name;
        }
    }
}

File: examples/dgs-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/dgsspringboot/datafetcher/AdditionalDataFetcher.kt
package org.neo4j.graphql.examples.dgsspringboot.datafetcher

import com.netflix.graphql.dgs.DgsComponent
import com.netflix.graphql.dgs.DgsData
import graphql.schema.DataFetchingEnvironment
import org.neo4j.graphql.examples.dgsspringboot.types.DgsConstants
import org.neo4j.graphql.examples.dgsspringboot.types.types.JavaData
import java.util.*


@DgsComponent
class AdditionalDataFetcher {
    @DgsData(parentType = DgsConstants.MOVIE.TYPE_NAME, field = DgsConstants.MOVIE.Bar)
    fun bar(): String {
        return "foo"
    }

    @DgsData(parentType = DgsConstants.MOVIE.TYPE_NAME, field = DgsConstants.MOVIE.JavaData)
    fun javaData(env: DataFetchingEnvironment): List<JavaData> {
        val title = env.getSource<Map<String, *>>()?.get("title")
        return Collections.singletonList(JavaData("test $title"))
    }

    @DgsData(parentType = DgsConstants.QUERY.TYPE_NAME, field = DgsConstants.QUERY.Other)
    fun other(): String {
        return "other"
    }
}
